### PR TITLE
Create modular windows

### PR DESCRIPTION
**Pull request by Raycoms**
# Changes proposed in this pull request:
- Create modular windows
- Finish modularization of filterable lists
-

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    public AbstractModuleWindow(final IBuildingView building, final String res)
+    {
+        super(res);
+        final Random random = new Random();
+        int offset = 0;
+        if (building.getAllModuleViews().size() > 0)
+        {
+            final ButtonImage image = new ButtonImage();
+            image.setImage("minecolonies:textures/gui/modules/tab_side" + (random.nextInt(3) + 1) + ".png");
+            image.setPosition(-20, 10 + offset);
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java:36]: stop using negative positions

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+            offset += image.getHeight() + 2;
+
+            this.addChild(image);
+            this.addChild(iconImage);
+
+            PaneBuilders.tooltipBuilder().hoverPane(iconImage).build().setText(new TranslationTextComponent("com.minecolonies.coremod.gui.maintab"));
+        }
+
+        for (IBuildingModuleView view : building.getAllModuleViews())
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java:55]: this seems to be in single column, using list might be better, you can hide the scrollbar if needed (needs bo change tho)

File: src/main/java/com/minecolonies/coremod/client/gui/ItemListModuleWindow.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    {
+        updateResources();
+    }
+
+    @Override
+    public boolean onKeyTyped(final char ch, final int key)
+    {
+        final boolean result = super.onKeyTyped(ch, key);
+        filter = window.findPaneOfTypeByID(INPUT_FILTER, TextField.class).getText();
+        updateResources();
```
Comment (Nightenom) [inline src/main/java/com/minecolonies/coremod/client/gui/ItemListModuleWindow.java]: only update if filter changes, otherwise any key type triggers update

**Review by Nightenom — DISMISSED** [review]

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java (hunk 2 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java:55]: For now we want this to be a single column, and if we have more one day, extend it to extend the other side on the other side of the UI

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java (hunk 1 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java:36]: So, I have to increase the overall window size then?

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java (hunk 2 above)
Comment (Nightenom) [reply src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java:55]: then create a proper handler

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java (hunk 1 above)
Comment (Nightenom) [reply src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java:36]: yep

File: src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java (hunk 2 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/client/gui/AbstractModuleWindow.java:55]: This is the proper place for this.

File: src/api/java/com/minecolonies/api/colony/buildings/modules/IBuildingModuleView.java [hunk 4]
```diff
@@ -21,4 +22,22 @@
      * @return this module itself.
      */
     IBuildingModuleView setBuildingView(final IBuildingView buildingView);
+
+    /**
+     * Get the matching window for the module.
+     * @return the window.
+     */
+    Window getWindow();
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/colony/buildings/modules/IBuildingModuleView.java:33]: You will need an @OnlyIn here or on the view.....
Or optionally a Supplier of a Supplier structure....

This won't work on a dedicated server, or at least is very dangerous.

**Review by marchermans — CHANGES_REQUESTED** [review]
Look good, i have some sidedness concerns with the Window being exposed this way.

File: src/api/java/com/minecolonies/api/colony/buildings/modules/IBuildingModuleView.java (hunk 4 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/colony/buildings/modules/IBuildingModuleView.java:33]: Aren't views always only client side?

Comment (someaddons) [reply src/api/java/com/minecolonies/api/colony/buildings/modules/IBuildingModuleView.java:33]: might be not due to the way we're constructing buildings now in our buildings entry right

Comment (uecasm) [reply src/api/java/com/minecolonies/api/colony/buildings/modules/IBuildingModuleView.java:33]: It'd probably be a good idea to have client-side-only modules in a different package from the server-side-only modules, at least. It's easier to spot when something is improperly importing across sides when they're segregated.

Comment (Raycoms) [reply src/api/java/com/minecolonies/api/colony/buildings/modules/IBuildingModuleView.java:33]: We're only giving it a producer in the entry, not an actual view. But I agree with @uecasm I'll make it a seperate folder

Comment (Raycoms) [reply src/api/java/com/minecolonies/api/colony/buildings/modules/IBuildingModuleView.java:33]: I double checked, this works fine in SMP too. But I moved it into seperate folders

**Pull request merged**